broadinstitute/single_cell_portal_core
# Fast validation of AnnData rows (SCP-5719)

This extends up-front checks to cover row-level data in an AnnData file. It makes authoring smoother, and thus faster.

### Impact
[Previously](https://github.com/broadinstitute/single_cell_portal_core/pull/2118), new AnnData validation caught errors in headers.  Fast validation for headers is very convenient, but superficial.  Most validation errors occur in the body of the AnnData file.

Now, deeper checks validate some _content_ of rows in an AnnData, for both local and remote HDF5 files.  A new basic rule examines the format of ontology IDs for rows of required metadata.  It covers up to [23% of file validation errors](https://mixpanel.com/project/2120588/view/19059/app/boards#id=734187&editor-card-id=%22report-28093580%22) in classic upload, so starting AnnData row validation with this relatively simple rule seems worthwhile.  

### Video

Here's how it looks!

https://github.com/user-attachments/assets/7f35cc3c-a3bd-4ea7-ba98-4ddec63cee0a

### Test
A new automated test verifies behavior.  Optional manual test:

1. Ensure Vite frontend service worker cache is disabled
2. Go to study "AnnData CSFV"
3. Select AnnData upload UX
4. In AnnData tab, click "Choose file"
5. In Finder, go to single_cell_portal_core/test/test_data/
6. Select anndata_test_invalid_disease.h5ad
7. Confirm error message is shown
8. Click "Delete" to remove errant file, then click "Choose file"
9. In same directory as before, select anndata_test.h5ad
10. Confirm no error message is shown, and "Save" button is enabled
11. Upload anndata_test_invalid_disease.h5ad and to your study's GCP bucket
12. Click "Use bucket path"
13. Enter "anndata_test_invalid_disease.h5ad" in text input
14. Click off input to blur / change browser focus
15. Confirm error message is shown is disabled

This satisfies SCP-5719.
